# Optimized Whisper on AWS Neuron (Trainium2 / Inferentia2)

This notebook runs OpenAI's [Whisper large-v3-turbo](https://huggingface.co/openai/whisper-large-v3-turbo) on AWS Neuron instances using [NxD Inference](https://github.com/aws-neuron/neuronx-distributed-inference) with four optimizations:

1. **Cross-attention K/V cache** — eliminates redundant K/V projections (19.7B FLOPs/token) and audio transfer (3.84MB/token) during decode
2. **Fused QKV projections** — replaces 3 separate Q/K/V linear layers with a single fused matmul in self-attention
3. **NKI flash attention** — replaces matmul-based attention with hardware-accelerated flash attention in all 32 encoder layers
4. **NKI fused Conv1D+GELU** — replaces separate Conv1D and GELU ops with a single fused NKI kernel in the encoder frontend

**Supported instances:** trn2.3xlarge, inf2.xlarge (or any Neuron instance with ≥1 NeuronCore)

**Model:** whisper-large-v3-turbo (809M params, 32 encoder layers, 4 decoder layers) at TP=1, FP16

## 1. Setup

Activate the pre-installed PyTorch Neuron environment and install dependencies.

```bash
source /opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/bin/activate
```

In [1]:
!pip install -q openai-whisper soundfile librosa gtts jiwer 2>/dev/null
!sudo add-apt-repository -y universe > /dev/null 2>&1 && sudo apt-get update -qq > /dev/null 2>&1 && sudo apt-get install -y -qq ffmpeg > /dev/null 2>&1
!which ffmpeg

/usr/bin/ffmpeg


## 2. Install Optimized NxDI

Clone our optimized fork of neuronx-distributed-inference and install it over the system package.
This replaces only the Whisper model files — all other NxDI models are unaffected.

**Note:** If running via `run_notebook.sh`, these packages are pre-installed. The cells below
detect this and skip the install if already present.

In [2]:
import os
import subprocess
import sys

NXDI_BRANCH = "fix/whisper-all-optimizations"
NXDI_REPO = "https://github.com/jimburtoft/neuronx-distributed-inference.git"
NXDI_CLONE_DIR = "/tmp/nxdi-optimized"

# Clone the optimized branch
if not os.path.exists(NXDI_CLONE_DIR):
    subprocess.run(
        ["git", "clone", "--branch", NXDI_BRANCH, "--single-branch", "--depth", "1", NXDI_REPO, NXDI_CLONE_DIR],
        check=True,
    )
    print(f"Cloned {NXDI_BRANCH} to {NXDI_CLONE_DIR}")
else:
    print(f"Already cloned at {NXDI_CLONE_DIR}")

# Install over the system NxDI package (if not already installed from run_notebook.sh)
try:
    import neuronx_distributed_inference
    print(f"NxDI already available: {neuronx_distributed_inference.__file__}")
except ImportError:
    subprocess.run([sys.executable, "-m", "pip", "install", "-q", "--no-deps", "--force-reinstall", NXDI_CLONE_DIR], check=True)
    print("Optimized NxDI installed — please restart kernel if running interactively")

Already cloned at /tmp/nxdi-optimized
NxDI already available: /opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed_inference/__init__.py


## 3. Install NKI Conv1D kernel (optional)

The DLAMI includes `nkilib` but may lack the newer `conv1d` kernel. We copy it
from the nki-library repo. If unavailable, the notebook falls back to PyTorch Conv1D.

In [3]:
import os, subprocess, sys

NKI_LIB_DIR = "/tmp/nki-library"

try:
    from nkilib.experimental.conv.conv1d import conv1d as nki_conv1d
    print("NKI Conv1D kernel already available")
except ImportError:
    # Copy conv1d.py from nki-library repo into the installed nkilib package
    try:
        import nkilib.experimental.conv
        conv_dir = os.path.dirname(nkilib.experimental.conv.__file__)
        if not os.path.exists(NKI_LIB_DIR):
            subprocess.run(["git", "clone", "--depth", "1",
                "https://github.com/aws-neuron/nki-library.git", NKI_LIB_DIR],
                check=True, capture_output=True)
        src = os.path.join(NKI_LIB_DIR, "src", "nkilib_src", "nkilib", "experimental", "conv", "conv1d.py")
        dst = os.path.join(conv_dir, "conv1d.py")
        import shutil
        shutil.copy2(src, dst)
        from nkilib.experimental.conv.conv1d import conv1d as nki_conv1d
        print(f"NKI Conv1D kernel installed to {dst}")
    except Exception as e:
        print(f"NKI Conv1D not available ({e}) — will use PyTorch fallback")

NKI Conv1D not available (No module named 'nki') — will use PyTorch fallback


## 4. Detect Hardware

Auto-detect whether we're on Trainium or Inferentia and configure accordingly.

In [4]:
import time
import json
import torch

os.environ["NEURON_RT_NUM_CORES"] = "1"

# NKI conv1d kernel needs platform hint during compilation
result = subprocess.run(["neuron-ls"], capture_output=True, text=True)
neuron_ls_output = result.stdout
print(neuron_ls_output)

if "trn2" in neuron_ls_output.lower():
    PLATFORM = "trn2"
    os.environ["NEURON_PLATFORM_TARGET_OVERRIDE"] = "trn2"
elif "trn1" in neuron_ls_output.lower():
    PLATFORM = "trn1"
elif "inf2" in neuron_ls_output.lower() or "inf1" in neuron_ls_output.lower():
    PLATFORM = "inf2"
else:
    # Fallback: check instance type from metadata
    try:
        token_result = subprocess.run(
            ["curl", "-s", "-X", "PUT", "http://169.254.169.254/latest/api/token",
             "-H", "X-aws-ec2-metadata-token-ttl-seconds: 60"],
            capture_output=True, text=True, timeout=2
        )
        token = token_result.stdout.strip()
        itype_result = subprocess.run(
            ["curl", "-s", "-H", f"X-aws-ec2-metadata-token: {token}",
             "http://169.254.169.254/latest/meta-data/instance-type"],
            capture_output=True, text=True, timeout=2
        )
        instance_type = itype_result.stdout.strip()
        if "trn2" in instance_type:
            PLATFORM = "trn2"
            os.environ["NEURON_PLATFORM_TARGET_OVERRIDE"] = "trn2"
        elif "trn1" in instance_type:
            PLATFORM = "trn1"
        elif "inf" in instance_type:
            PLATFORM = "inf2"
        else:
            PLATFORM = "unknown"
    except Exception:
        PLATFORM = "unknown"

print(f"\nDetected platform: {PLATFORM}")

instance-type: inf2.xlarge
instance-id: i-0d2acdecd24de369a
+--------+--------+----------+--------+--------------+----------+------+
| NEURON | NEURON |  NEURON  | NEURON |     PCI      |   CPU    | NUMA |
| DEVICE | CORES  | CORE IDS | MEMORY |     BDF      | AFFINITY | NODE |
+--------+--------+----------+--------+--------------+----------+------+
| 0      | 2      | 0-1      | 32 GB  | 0000:00:1f.0 | 0-3      | -1   |
+--------+--------+----------+--------+--------------+----------+------+


Detected platform: inf2


In [5]:
from neuronx_distributed_inference.models.config import NeuronConfig
from neuronx_distributed_inference.models.whisper.modeling_whisper import (
    WhisperInferenceConfig,
    NeuronApplicationWhisper,
)
from neuronx_distributed_inference.utils.hf_adapter import load_pretrained_config

/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/parallel_layers/layers.py:16: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  from .mappings import (
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/parallel_layers/layers.py:16: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  from .mappings import (
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/parallel_layers/layers.py:16: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  from .mappings import (


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/modules/moe/blockwise.py:74: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  component, error = import_nki(config)
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/modules/moe/blockwise.py:74: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  component, error = import_nki(config)
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/modules/moe/blockwise.py:74: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  component, error = import_nki(config)
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/modules/moe/blockwise.py:74: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  component, error = import_nki(config)
/opt/aws_neuronx

/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/modules/moe/moe_fused_tkg.py:49: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  component, error = import_nki(config)


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/modules/moe/moe_fused_tkg.py:49: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  component, error = import_nki(config)
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/modules/moe/moe_fused_tkg.py:49: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  component, error = import_nki(config)


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed_inference/modules/attention/utils.py:13: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  from neuronx_distributed_inference.modules.custom_calls import neuron_cumsum
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed_inference/modules/lora_serving/lora_checkpoint.py:9: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  from neuronx_distributed_inference.modules.attention.gqa import replicate_kv
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed_inference/modules/lora_serving/lora_checkpoint.py:9: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  from neuronx_distributed_inference.modules.attention.gqa import replicate_kv
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuron

## 5. Download Model

In [6]:
MODEL_PATH = "/tmp/whisper-large-v3-turbo/"
COMPILED_MODEL_PATH = f"/tmp/whisper-large-v3-turbo-neuron-{PLATFORM}/"

if not os.path.exists(os.path.join(MODEL_PATH, "config.json")):
    from huggingface_hub import snapshot_download
    snapshot_download(
        repo_id="openai/whisper-large-v3-turbo",
        local_dir=MODEL_PATH,
        ignore_patterns=["*.onnx", "*.tflite", "flax_model*", "tf_model*"],
    )
    print(f"Downloaded to {MODEL_PATH}")
else:
    print(f"Model already exists at {MODEL_PATH}")

Model already exists at /tmp/whisper-large-v3-turbo/


## 6. Compile for Neuron

NxDI compiles three separate graphs:
- **Encoder** — processes the 30-second mel spectrogram window
- **Decoder Prefill** — processes all prompt tokens at once, caches encoder K/V
- **Decoder Decode** — generates one token at a time using KV cache

Compilation is a one-time cost. The compiled model is platform-specific (trn2 vs inf2).

In [7]:
neuron_config = NeuronConfig(
    batch_size=1,
    torch_dtype=torch.float16,
    tp_degree=1,
)
inference_config = WhisperInferenceConfig(
    neuron_config,
    load_config=load_pretrained_config(MODEL_PATH),
)

if not os.path.exists(COMPILED_MODEL_PATH):
    print(f"Compiling for {PLATFORM} (one-time cost, ~60-120s)...")
    t0 = time.time()
    model = NeuronApplicationWhisper(MODEL_PATH, config=inference_config)
    model.compile(COMPILED_MODEL_PATH)
    compile_time = time.time() - t0
    print(f"Compilation complete in {compile_time:.1f}s")
else:
    compile_time = None
    print(f"Using cached compiled model at {COMPILED_MODEL_PATH}")

Compiling for inf2 (one-time cost, ~60-120s)...


INFO:Neuron:Saving the neuron_config to /tmp/whisper-large-v3-turbo-neuron-inf2/encoder/


INFO:Neuron:Generating HLOs for the following models: ['Encoder']


[2026-03-11 19:56:53.303: I neuronx_distributed/parallel_layers/parallel_state.py:638] > initializing tensor model parallel with size 1


[2026-03-11 19:56:53.304: I neuronx_distributed/parallel_layers/parallel_state.py:639] > initializing pipeline model parallel with size 1


[2026-03-11 19:56:53.304: I neuronx_distributed/parallel_layers/parallel_state.py:640] > initializing context model parallel with size 1


[2026-03-11 19:56:53.305: I neuronx_distributed/parallel_layers/parallel_state.py:641] > initializing data parallel with size 1


[2026-03-11 19:56:53.305: I neuronx_distributed/parallel_layers/parallel_state.py:642] > initializing world size to 1


[2026-03-11 19:56:53.306: I neuronx_distributed/parallel_layers/parallel_state.py:387] [rank_0_pp-1_tp-1_dp-1_cp-1] Chosen Logic for replica groups ret_logic=<PG_Group_Logic.LOGIC1: (<function ascending_ring_PG_group at 0x70a8387a99e0>, 'Ascending Ring PG Group')>


[2026-03-11 19:56:53.306: I neuronx_distributed/parallel_layers/parallel_state.py:666] [rank_0_pp-1_tp-1_dp-1_cp-1] tp_groups: replica_groups.tp_groups=[[0]]


[2026-03-11 19:56:53.307: I neuronx_distributed/parallel_layers/parallel_state.py:667] [rank_0_pp-1_tp-1_dp-1_cp-1] dp_groups: replica_groups.dp_groups=[[0]]


[2026-03-11 19:56:53.307: I neuronx_distributed/parallel_layers/parallel_state.py:668] [rank_0_pp-1_tp-1_dp-1_cp-1] pp_groups: replica_groups.pp_groups=[[0]]


[2026-03-11 19:56:53.308: I neuronx_distributed/parallel_layers/parallel_state.py:669] [rank_0_pp-1_tp-1_dp-1_cp-1] cp_groups: replica_groups.cp_groups=[[0]]


[2026-03-11 19:56:53.308: I neuronx_distributed/parallel_layers/parallel_state.py:670] [rank_0_pp-1_tp-1_dp-1_cp-1] ep_model_groups: replica_groups.ep_model_groups=[[0]]


[2026-03-11 19:56:53.309: I neuronx_distributed/parallel_layers/parallel_state.py:671] [rank_0_pp-1_tp-1_dp-1_cp-1] ep_data_groups: replica_groups.ep_data_groups=[[0]]


INFO:Neuron:Generating 1 hlos for key: Encoder


INFO:Neuron:Minimal metadata will be added to HLO


INFO:Neuron:Started loading module Encoder


INFO:Neuron:Finished loading module Encoder in 0.07356762886047363 seconds


INFO:Neuron:generating HLO: Encoder, input example shape = torch.Size([1, 128, 3000])


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/parallel_layers/layers.py:532: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with torch.cuda.amp.autocast(enabled=False):
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed_inference/models/whisper/modeling_whisper.py:157: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  output = scaled_dot_product_attention_kernel(
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/parallel_layers/layers.py:532: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with torch.cuda.amp.autocast(enabled=False):
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed_inference/models/whisper/m

/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed_inference/models/whisper/modeling_whisper.py:157: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  output = scaled_dot_product_attention_kernel(
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/parallel_layers/layers.py:532: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with torch.cuda.amp.autocast(enabled=False):
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed_inference/models/whisper/modeling_whisper.py:157: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  output = scaled_dot_product_attention_kernel(
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/parallel_layers/layers.py:532: FutureWarning: `torc

Neuron NKI - Kernel call: attention_isa_kernel(q = Tensor(shape: (20, 64, 1500), dtype: float16), k = Tensor(shape: (20, 64, 1500), dtype: float16), v = Tensor(shape: (20, 1500, 64), dtype: float16), scale = 1.0, out = Tensor(shape: (20, 64, 1500), dtype: float16), kernel_name = AttentionMMSoftmaxMMWithoutSwap, use_dma_transpose = True, global_n_tiles = -1, tile_i = None, strided_q_slicing = False, use_psum_vec = False, do_out_tp = False, sink = None, sliding_window = 0)


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed_inference/models/whisper/modeling_whisper.py:157: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  output = scaled_dot_product_attention_kernel(
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/parallel_layers/layers.py:532: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with torch.cuda.amp.autocast(enabled=False):
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed_inference/models/whisper/modeling_whisper.py:157: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  output = scaled_dot_product_attention_kernel(
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/parallel_layers/layers.py:532: FutureWarning: `torc

/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed_inference/models/whisper/modeling_whisper.py:157: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  output = scaled_dot_product_attention_kernel(
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/parallel_layers/layers.py:532: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with torch.cuda.amp.autocast(enabled=False):
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed_inference/models/whisper/modeling_whisper.py:157: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  output = scaled_dot_product_attention_kernel(
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/parallel_layers/layers.py:532: FutureWarning: `torc

/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed_inference/models/whisper/modeling_whisper.py:157: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  output = scaled_dot_product_attention_kernel(
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/parallel_layers/layers.py:532: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with torch.cuda.amp.autocast(enabled=False):
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed_inference/models/whisper/modeling_whisper.py:157: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  output = scaled_dot_product_attention_kernel(
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/parallel_layers/layers.py:532: FutureWarning: `torc

/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed_inference/models/whisper/modeling_whisper.py:157: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  output = scaled_dot_product_attention_kernel(
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/parallel_layers/layers.py:532: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with torch.cuda.amp.autocast(enabled=False):
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed_inference/models/whisper/modeling_whisper.py:157: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  output = scaled_dot_product_attention_kernel(
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/parallel_layers/layers.py:532: FutureWarning: `torc

/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed_inference/models/whisper/modeling_whisper.py:157: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  output = scaled_dot_product_attention_kernel(
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/parallel_layers/layers.py:532: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with torch.cuda.amp.autocast(enabled=False):
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed_inference/models/whisper/modeling_whisper.py:157: DeprecationWarning: torch_neuronx.nki_jit is deprecated, use nki.jit instead.
  output = scaled_dot_product_attention_kernel(
/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/parallel_layers/layers.py:532: FutureWarning: `torc

INFO:Neuron:Finished generating HLO for Encoder in 1.0622494220733643 seconds, input example shape = torch.Size([1, 128, 3000])


INFO:Neuron:Generated all HLOs in 1.1894903182983398 seconds


INFO:Neuron:Can't find a priority model, skip marking weights


INFO:Neuron:Can't find a priority model, skip optimizing weight layout for other HLOs


INFO:Neuron:Starting compilation for all HLOs


INFO:Neuron:Neuron compiler flags: --model-type=transformer --tensorizer-options='--enable-ccop-compute-overlap --cc-pipeline-tiling-factor=2' --internal-hlo2tensorizer-options='--verify-hlo=true'  --auto-cast=none  -O1  --verbose=35 --logfile=/tmp/nxd_model/Encoder/_tp0_bk0/log-neuron-cc.txt


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/libneuronxla/neuron_cc_wrapper.py:246: SyntaxWarning: str format compiler_flags is discouraged as its handling involves repeated joining and splitting, which can easily make mistakes if something is quoted or escaped. Use list[str] instead. Refer to documentation of the Python subprocess module for details.
  warnings.warn(SyntaxWarning(


.

Roundtrip constructed a transpose sequence [rounds: 1; efficiency: 77]:
  dve_j_optimized   : Fix prefix (0, 1) and permute (2,) with (3, 4) / latency=13,460; shape=(10, 128, 128, 1, 3); dtype_size=2

Roundtrip constructed a transpose sequence [rounds: 1; efficiency: 60]:
  dve_j_optimized   : Fix prefix (0, 1, 2) and permute (3,) with (4, 5) / latency=134,603; shape=(10, 128, 10, 128, 1, 3); dtype_size=2

Neuron NKI - Kernel call: tiled_dve_transpose_10(in_tensor = Tensor(shape: (10, 128, 128, 1, 3), dtype: float16), in_shape = [10, 128, 128, 1, 3], permutation = [0, 1, 3, 4, 2])
Neuron NKI - Kernel call: tiled_dve_transpose_10(in_tensor = Tensor(shape: (10, 128, 10, 128, 1, 3), dtype: float16), in_shape = [10, 128, 10, 128, 1, 3], permutation = [0, 1, 2, 4, 5, 3])



Compiler status PASS
2026-03-11 19:57:04.000545:  19262  [INFO]: Compilation Successfully Completed for model.MODULE_58048f956b8c44bc5b5b+c9c9fa2d.hlo_module.pb


INFO:Neuron:Finished Compilation for all HLOs in 10.056217193603516 seconds


INFO:Neuron:Can't find a priority model, falling back to the existing weight layout


INFO:Neuron:Finished building model in 11.457454681396484 seconds


INFO:Neuron:SKIPPING pre-sharding the checkpoints. The checkpoints will be sharded during load time.


INFO:Neuron:Saving the neuron_config to /tmp/whisper-large-v3-turbo-neuron-inf2/decoder/


INFO:Neuron:Generating HLOs for the following models: ['DecoderPrefill', 'DecoderDecode']


[2026-03-11 19:57:04.813: I neuronx_distributed/parallel_layers/parallel_state.py:638] > initializing tensor model parallel with size 1


[2026-03-11 19:57:04.813: I neuronx_distributed/parallel_layers/parallel_state.py:639] > initializing pipeline model parallel with size 1


[2026-03-11 19:57:04.814: I neuronx_distributed/parallel_layers/parallel_state.py:640] > initializing context model parallel with size 1


[2026-03-11 19:57:04.814: I neuronx_distributed/parallel_layers/parallel_state.py:641] > initializing data parallel with size 1


[2026-03-11 19:57:04.814: I neuronx_distributed/parallel_layers/parallel_state.py:642] > initializing world size to 1


[2026-03-11 19:57:04.815: I neuronx_distributed/parallel_layers/parallel_state.py:387] [rank_0_pp-1_tp-1_dp-1_cp-1] Chosen Logic for replica groups ret_logic=<PG_Group_Logic.LOGIC1: (<function ascending_ring_PG_group at 0x70a8387a99e0>, 'Ascending Ring PG Group')>


[2026-03-11 19:57:04.816: I neuronx_distributed/parallel_layers/parallel_state.py:666] [rank_0_pp-1_tp-1_dp-1_cp-1] tp_groups: replica_groups.tp_groups=[[0]]


[2026-03-11 19:57:04.816: I neuronx_distributed/parallel_layers/parallel_state.py:667] [rank_0_pp-1_tp-1_dp-1_cp-1] dp_groups: replica_groups.dp_groups=[[0]]


[2026-03-11 19:57:04.816: I neuronx_distributed/parallel_layers/parallel_state.py:668] [rank_0_pp-1_tp-1_dp-1_cp-1] pp_groups: replica_groups.pp_groups=[[0]]


[2026-03-11 19:57:04.817: I neuronx_distributed/parallel_layers/parallel_state.py:669] [rank_0_pp-1_tp-1_dp-1_cp-1] cp_groups: replica_groups.cp_groups=[[0]]


[2026-03-11 19:57:04.817: I neuronx_distributed/parallel_layers/parallel_state.py:670] [rank_0_pp-1_tp-1_dp-1_cp-1] ep_model_groups: replica_groups.ep_model_groups=[[0]]


[2026-03-11 19:57:04.818: I neuronx_distributed/parallel_layers/parallel_state.py:671] [rank_0_pp-1_tp-1_dp-1_cp-1] ep_data_groups: replica_groups.ep_data_groups=[[0]]


INFO:Neuron:Generating 1 hlos for key: DecoderPrefill


INFO:Neuron:Minimal metadata will be added to HLO


INFO:Neuron:Started loading module DecoderPrefill


INFO:Neuron:Finished loading module DecoderPrefill in 0.3630366325378418 seconds


INFO:Neuron:generating HLO: DecoderPrefill, input example shape = torch.Size([1, 448])


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/parallel_layers/layers.py:532: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with torch.cuda.amp.autocast(enabled=False):


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/torch_neuronx/xla_impl/hlo_conversion.py:470: UserWarning: Received an input tensor that was unused or used in a non-static way when traced so the tensor will be ignored. (index=2, shape=torch.Size([1]), dtype=torch.int32). The non-static usage could happen when the traced function expects the input tensor's shape to change (i.e., using the shape to do index slicing), which is not allowed by inference trace expecting static input shapes.
  warnings.warn(


INFO:Neuron:Finished generating HLO for DecoderPrefill in 0.26863622665405273 seconds, input example shape = torch.Size([1, 448])


INFO:Neuron:Generating 1 hlos for key: DecoderDecode


INFO:Neuron:Minimal metadata will be added to HLO


INFO:Neuron:Started loading module DecoderDecode


INFO:Neuron:Finished loading module DecoderDecode in 0.3662993907928467 seconds


INFO:Neuron:generating HLO: DecoderDecode, input example shape = torch.Size([1, 1])


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/torch_neuronx/xla_impl/hlo_conversion.py:470: UserWarning: Received an input tensor that was unused or used in a non-static way when traced so the tensor will be ignored. (index=1, shape=torch.Size([1, 1, 1280]), dtype=torch.float16). The non-static usage could happen when the traced function expects the input tensor's shape to change (i.e., using the shape to do index slicing), which is not allowed by inference trace expecting static input shapes.
  warnings.warn(


INFO:Neuron:Finished generating HLO for DecoderDecode in 0.24385309219360352 seconds, input example shape = torch.Size([1, 1])


INFO:Neuron:Generated all HLOs in 1.3055095672607422 seconds


INFO:Neuron:Can't find a priority model, skip marking weights


INFO:Neuron:Can't find a priority model, skip optimizing weight layout for other HLOs


INFO:Neuron:Starting compilation for all HLOs


INFO:Neuron:Neuron compiler flags: --model-type=transformer --tensorizer-options='--enable-ccop-compute-overlap --cc-pipeline-tiling-factor=2' --internal-hlo2tensorizer-options='--verify-hlo=true'  --auto-cast=none  -O1  --verbose=35 --logfile=/tmp/nxd_model/DecoderPrefill/_tp0_bk0/log-neuron-cc.txt


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/libneuronxla/neuron_cc_wrapper.py:246: SyntaxWarning: str format compiler_flags is discouraged as its handling involves repeated joining and splitting, which can easily make mistakes if something is quoted or escaped. Use list[str] instead. Refer to documentation of the Python subprocess module for details.
  warnings.warn(SyntaxWarning(
INFO:Neuron:Neuron compiler flags: --model-type=transformer --tensorizer-options='--enable-ccop-compute-overlap --cc-pipeline-tiling-factor=2' --internal-hlo2tensorizer-options='--verify-hlo=true'  --auto-cast=none  -O1  --verbose=35 --logfile=/tmp/nxd_model/DecoderDecode/_tp0_bk0/log-neuron-cc.txt


..


Compiler status PASS
2026-03-11 19:57:22.000367:  19262  [INFO]: Compilation Successfully Completed for model.MODULE_40f4248e0de49afc64c4+7ef8efc5.hlo_module.pb



Compiler status PASS
2026-03-11 19:57:25.000959:  19262  [INFO]: Compilation Successfully Completed for model.MODULE_6a1f7ad60f76d75316ce+f69ccc06.hlo_module.pb


INFO:Neuron:Finished Compilation for all HLOs in 19.84364342689514 seconds


INFO:Neuron:Can't find a priority model, falling back to the existing weight layout


INFO:Neuron:Finished building model in 22.40447688102722 seconds


INFO:Neuron:SKIPPING pre-sharding the checkpoints. The checkpoints will be sharded during load time.


Compilation complete in 38.1s


## 7. Load Compiled Model

In [8]:
t0 = time.time()
model = NeuronApplicationWhisper(COMPILED_MODEL_PATH, config=inference_config)
model.load(COMPILED_MODEL_PATH)
load_time = time.time() - t0
print(f"Model loaded in {load_time:.1f}s")

INFO:Neuron:Sharding weights on load...


INFO:Neuron:Sharding weights for ranks: 0...0


[2026-03-11 19:57:31.455: I neuronx_distributed/parallel_layers/parallel_state.py:638] > initializing tensor model parallel with size 1


[2026-03-11 19:57:31.455: I neuronx_distributed/parallel_layers/parallel_state.py:639] > initializing pipeline model parallel with size 1


[2026-03-11 19:57:31.455: I neuronx_distributed/parallel_layers/parallel_state.py:640] > initializing context model parallel with size 1


[2026-03-11 19:57:31.456: I neuronx_distributed/parallel_layers/parallel_state.py:641] > initializing data parallel with size 1


[2026-03-11 19:57:31.457: I neuronx_distributed/parallel_layers/parallel_state.py:642] > initializing world size to 1


[2026-03-11 19:57:31.457: I neuronx_distributed/parallel_layers/parallel_state.py:387] [rank_0_pp-1_tp-1_dp-1_cp-1] Chosen Logic for replica groups ret_logic=<PG_Group_Logic.LOGIC1: (<function ascending_ring_PG_group at 0x70a8387a99e0>, 'Ascending Ring PG Group')>


[2026-03-11 19:57:31.458: I neuronx_distributed/parallel_layers/parallel_state.py:666] [rank_0_pp-1_tp-1_dp-1_cp-1] tp_groups: replica_groups.tp_groups=[[0]]


[2026-03-11 19:57:31.459: I neuronx_distributed/parallel_layers/parallel_state.py:667] [rank_0_pp-1_tp-1_dp-1_cp-1] dp_groups: replica_groups.dp_groups=[[0]]


[2026-03-11 19:57:31.459: I neuronx_distributed/parallel_layers/parallel_state.py:668] [rank_0_pp-1_tp-1_dp-1_cp-1] pp_groups: replica_groups.pp_groups=[[0]]


[2026-03-11 19:57:31.460: I neuronx_distributed/parallel_layers/parallel_state.py:669] [rank_0_pp-1_tp-1_dp-1_cp-1] cp_groups: replica_groups.cp_groups=[[0]]


[2026-03-11 19:57:31.460: I neuronx_distributed/parallel_layers/parallel_state.py:670] [rank_0_pp-1_tp-1_dp-1_cp-1] ep_model_groups: replica_groups.ep_model_groups=[[0]]


[2026-03-11 19:57:31.460: I neuronx_distributed/parallel_layers/parallel_state.py:671] [rank_0_pp-1_tp-1_dp-1_cp-1] ep_data_groups: replica_groups.ep_data_groups=[[0]]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/neuronx_distributed/trace/trace.py:642: UserWarning: Removing redundant keys from checkpoint: []
  warnings.warn(f"Removing redundant keys from checkpoint: {keys_to_delete}")


INFO:Neuron:Done Sharding weights in 0.5370318360000965


INFO:Neuron:Finished weights loading in 6.5832811400005085 seconds


INFO:Neuron:Warming up the model.


INFO:Neuron:Warmup completed in 0.08985638618469238 seconds.


INFO:Neuron:Sharding weights on load...


INFO:Neuron:Sharding weights for ranks: 0...0


[2026-03-11 19:57:38.179: I neuronx_distributed/parallel_layers/parallel_state.py:638] > initializing tensor model parallel with size 1


[2026-03-11 19:57:38.180: I neuronx_distributed/parallel_layers/parallel_state.py:639] > initializing pipeline model parallel with size 1


[2026-03-11 19:57:38.180: I neuronx_distributed/parallel_layers/parallel_state.py:640] > initializing context model parallel with size 1


[2026-03-11 19:57:38.181: I neuronx_distributed/parallel_layers/parallel_state.py:641] > initializing data parallel with size 1


[2026-03-11 19:57:38.181: I neuronx_distributed/parallel_layers/parallel_state.py:642] > initializing world size to 1


[2026-03-11 19:57:38.182: I neuronx_distributed/parallel_layers/parallel_state.py:387] [rank_0_pp-1_tp-1_dp-1_cp-1] Chosen Logic for replica groups ret_logic=<PG_Group_Logic.LOGIC1: (<function ascending_ring_PG_group at 0x70a8387a99e0>, 'Ascending Ring PG Group')>


[2026-03-11 19:57:38.182: I neuronx_distributed/parallel_layers/parallel_state.py:666] [rank_0_pp-1_tp-1_dp-1_cp-1] tp_groups: replica_groups.tp_groups=[[0]]


[2026-03-11 19:57:38.183: I neuronx_distributed/parallel_layers/parallel_state.py:667] [rank_0_pp-1_tp-1_dp-1_cp-1] dp_groups: replica_groups.dp_groups=[[0]]


[2026-03-11 19:57:38.183: I neuronx_distributed/parallel_layers/parallel_state.py:668] [rank_0_pp-1_tp-1_dp-1_cp-1] pp_groups: replica_groups.pp_groups=[[0]]


[2026-03-11 19:57:38.184: I neuronx_distributed/parallel_layers/parallel_state.py:669] [rank_0_pp-1_tp-1_dp-1_cp-1] cp_groups: replica_groups.cp_groups=[[0]]


[2026-03-11 19:57:38.185: I neuronx_distributed/parallel_layers/parallel_state.py:670] [rank_0_pp-1_tp-1_dp-1_cp-1] ep_model_groups: replica_groups.ep_model_groups=[[0]]


[2026-03-11 19:57:38.185: I neuronx_distributed/parallel_layers/parallel_state.py:671] [rank_0_pp-1_tp-1_dp-1_cp-1] ep_data_groups: replica_groups.ep_data_groups=[[0]]


INFO:Neuron:Done Sharding weights in 0.13614392599993153


INFO:Neuron:Finished weights loading in 0.3675366010002108 seconds


INFO:Neuron:Warming up the model.


INFO:Neuron:Warmup completed in 0.10592055320739746 seconds.


Model loaded in 11.4s


## 8. Generate Test Audio

Create three speech samples of different lengths using Google Text-to-Speech.

In [9]:
import shutil
import soundfile as sf
import numpy as np
from gtts import gTTS

FFMPEG = shutil.which("ffmpeg") or "/usr/bin/ffmpeg"
AUDIO_DIR = "/tmp/whisper_audio"
os.makedirs(AUDIO_DIR, exist_ok=True)

# Three test samples of increasing length
samples = {
    "short": (
        "The quick brown fox jumps over the lazy dog. "
        "This is a test of automatic speech recognition."
    ),
    "medium": (
        "Artificial intelligence has transformed the way we interact with technology. "
        "Machine learning models can now understand and generate human language with remarkable accuracy. "
        "Speech recognition systems like Whisper have made it possible to transcribe audio in real time, "
        "supporting dozens of languages and handling noisy environments with ease. "
        "The combination of hardware acceleration and optimized software makes these capabilities "
        "accessible at scale, enabling applications from live captioning to voice assistants."
    ),
    "long": (
        "The history of artificial intelligence is a fascinating journey through decades of research and discovery. "
        "In the nineteen fifties, pioneers like Alan Turing asked whether machines could think, "
        "laying the philosophical groundwork for an entirely new field of computer science. "
        "Early systems used symbolic reasoning and hand-crafted rules to solve narrowly defined problems. "
        "The nineteen eighties saw the rise of expert systems, which encoded human knowledge into decision trees. "
        "But it was the advent of deep learning in the twenty tens that truly revolutionized the field. "
        "Neural networks with millions of parameters could learn directly from data, "
        "outperforming traditional approaches on tasks from image classification to natural language processing. "
        "Today, large language models and multimodal systems represent the cutting edge, "
        "capable of understanding text, images, and speech simultaneously. "
        "Custom silicon like AWS Trainium and Inferentia accelerates these workloads, "
        "making advanced AI accessible and cost-effective for organizations of every size."
    ),
}

audio_files = {}
for name, text in samples.items():
    wav_path = os.path.join(AUDIO_DIR, f"{name}.wav")
    if not os.path.exists(wav_path):
        mp3_path = os.path.join(AUDIO_DIR, f"{name}.mp3")
        tts = gTTS(text, lang="en")
        tts.save(mp3_path)
        subprocess.run(
            [FFMPEG, "-y", "-i", mp3_path, "-ar", "16000", "-ac", "1", wav_path],
            capture_output=True, check=True,
        )
    data, sr = sf.read(wav_path)
    duration = len(data) / sr
    audio_files[name] = {"path": wav_path, "duration": duration, "reference": text}
    print(f"{name:7s}: {duration:.1f}s — {wav_path}")

short  : 6.9s — /tmp/whisper_audio/short.wav


medium : 36.5s — /tmp/whisper_audio/medium.wav


long   : 76.7s — /tmp/whisper_audio/long.wav


## 9. Transcribe

Run transcription on all test samples. The NxDI model exposes the same `transcribe()` API as OpenAI Whisper.

In [10]:
# Warmup run
_ = model.transcribe(audio_files["short"]["path"], language="en", verbose=False)

for name, info in audio_files.items():
    result = model.transcribe(info["path"], language="en", verbose=False)
    info["neuron_text"] = result["text"].strip()
    print(f"\n--- {name} ({info['duration']:.1f}s) ---")
    print(f"Neuron: {info['neuron_text']}")

/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/691 [00:00<?, ?frames/s]

100%|██████████| 691/691 [00:00<00:00, 1251.37frames/s]

100%|██████████| 691/691 [00:00<00:00, 1249.77frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/691 [00:00<?, ?frames/s]

100%|██████████| 691/691 [00:00<00:00, 3130.48frames/s]

100%|██████████| 691/691 [00:00<00:00, 3120.74frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")



--- short (6.9s) ---
Neuron: The quick brown fox jumps over the lazy dog. This is a test of automatic speech recognition.


  0%|          | 0/3648 [00:00<?, ?frames/s]

 80%|███████▉  | 2910/3648 [00:00<00:00, 6787.60frames/s]

100%|██████████| 3648/3648 [00:02<00:00, 1009.43frames/s]

100%|██████████| 3648/3648 [00:02<00:00, 1267.37frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")



--- medium (36.5s) ---
Neuron: Artificial intelligence has transformed the way we interact with technology. Machine learning models can now understand and generate human language with remarkable accuracy. Speech recognition systems like Whisper have made it possible to transcribe audio in real time. Supporting dozens of languages and handling noisy environments with ease. The combination of hardware acceleration and optimized software makes these capabilities accessible at scale. Enabling applications from live captioning to voice assistance. Select acceptance


  0%|          | 0/7670 [00:00<?, ?frames/s]

 33%|███▎      | 2522/7670 [00:00<00:00, 6340.22frames/s]

 71%|███████▏  | 5484/7670 [00:00<00:00, 6433.08frames/s]

 99%|█████████▊| 7568/7670 [00:01<00:00, 6225.82frames/s]

100%|██████████| 7670/7670 [00:06<00:00, 1097.48frames/s]


--- long (76.7s) ---
Neuron: The history of artificial intelligence is a fascinating journey through decades of research and discovery. In the 1950s, pioneers like Alan Turing asked whether machines could think, laying the philosophical groundwork for an entirely new field of computer science. Early systems used symbolic reasoning and handcrafted rules to solve narrowly defined problems. The 1980s saw the rise of expert systems, which encoded human knowledge into decision trees. But it was the advent of deep learning in the 2010s that truly revolutionized the field. Neural networks with millions of parameters could learn directly from data, outperforming traditional approaches on tasks from image classification to natural language processing. Today, large language models and multimodal systems represent the cutting edge, capable of understanding text, images, and speech simultaneously. Custom silicon like AWS Tranium and Inferentia accelerates these workloads, making advanced AI acces

## 10. Benchmark

Measure latency across 10 runs per audio sample.

In [11]:
NUM_RUNS = 10

for name, info in audio_files.items():
    latencies = []
    for _ in range(NUM_RUNS):
        t0 = time.time()
        _ = model.transcribe(info["path"], language="en", verbose=False)
        latencies.append(time.time() - t0)
    info["neuron_avg_latency"] = np.mean(latencies)
    info["neuron_min_latency"] = min(latencies)
    info["neuron_rtf"] = info["duration"] / info["neuron_avg_latency"]

print(f"{'Audio':<8} {'Duration':>8} {'Avg Latency':>12} {'Min Latency':>12} {'RTF':>6}")
print("-" * 50)
for name, info in audio_files.items():
    print(f"{name:<8} {info['duration']:>7.1f}s {info['neuron_avg_latency']:>11.3f}s {info['neuron_min_latency']:>11.3f}s {info['neuron_rtf']:>5.1f}x")

/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/691 [00:00<?, ?frames/s]

100%|██████████| 691/691 [00:00<00:00, 3155.47frames/s]

100%|██████████| 691/691 [00:00<00:00, 3146.09frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/691 [00:00<?, ?frames/s]

100%|██████████| 691/691 [00:00<00:00, 3171.91frames/s]

100%|██████████| 691/691 [00:00<00:00, 3162.09frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/691 [00:00<?, ?frames/s]

100%|██████████| 691/691 [00:00<00:00, 3169.73frames/s]

100%|██████████| 691/691 [00:00<00:00, 3160.33frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/691 [00:00<?, ?frames/s]

100%|██████████| 691/691 [00:00<00:00, 3187.82frames/s]

100%|██████████| 691/691 [00:00<00:00, 3177.93frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/691 [00:00<?, ?frames/s]

100%|██████████| 691/691 [00:00<00:00, 3154.77frames/s]

100%|██████████| 691/691 [00:00<00:00, 3145.74frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/691 [00:00<?, ?frames/s]

100%|██████████| 691/691 [00:00<00:00, 3177.50frames/s]

100%|██████████| 691/691 [00:00<00:00, 3168.18frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/691 [00:00<?, ?frames/s]

100%|██████████| 691/691 [00:00<00:00, 3172.74frames/s]

100%|██████████| 691/691 [00:00<00:00, 3162.59frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/691 [00:00<?, ?frames/s]

100%|██████████| 691/691 [00:00<00:00, 3188.02frames/s]

100%|██████████| 691/691 [00:00<00:00, 3178.30frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/691 [00:00<?, ?frames/s]

100%|██████████| 691/691 [00:00<00:00, 3173.98frames/s]

100%|██████████| 691/691 [00:00<00:00, 3163.77frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/691 [00:00<?, ?frames/s]

100%|██████████| 691/691 [00:00<00:00, 3160.66frames/s]

100%|██████████| 691/691 [00:00<00:00, 3151.32frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/3648 [00:00<?, ?frames/s]

 80%|███████▉  | 2910/3648 [00:00<00:00, 6806.90frames/s]

100%|██████████| 3648/3648 [00:02<00:00, 1338.25frames/s]

100%|██████████| 3648/3648 [00:02<00:00, 1656.26frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/3648 [00:00<?, ?frames/s]

 80%|███████▉  | 2910/3648 [00:00<00:00, 6767.62frames/s]

100%|██████████| 3648/3648 [00:01<00:00, 1833.50frames/s]

100%|██████████| 3648/3648 [00:01<00:00, 2220.07frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/3648 [00:00<?, ?frames/s]

 80%|███████▉  | 2910/3648 [00:00<00:00, 6840.53frames/s]

100%|██████████| 3648/3648 [00:02<00:00, 1059.24frames/s]

100%|██████████| 3648/3648 [00:02<00:00, 1327.36frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/3648 [00:00<?, ?frames/s]

 80%|███████▉  | 2910/3648 [00:00<00:00, 6805.33frames/s]

100%|██████████| 3648/3648 [00:01<00:00, 1937.38frames/s]

100%|██████████| 3648/3648 [00:01<00:00, 2336.43frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/3648 [00:00<?, ?frames/s]

 80%|███████▉  | 2910/3648 [00:00<00:00, 6818.49frames/s]

100%|██████████| 3648/3648 [00:02<00:00, 978.81frames/s] 

100%|██████████| 3648/3648 [00:02<00:00, 1230.83frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/3648 [00:00<?, ?frames/s]

 80%|███████▉  | 2910/3648 [00:00<00:00, 6809.08frames/s]

100%|██████████| 3648/3648 [00:01<00:00, 1505.90frames/s]

100%|██████████| 3648/3648 [00:01<00:00, 1850.21frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/3648 [00:00<?, ?frames/s]

 80%|███████▉  | 2910/3648 [00:00<00:00, 6784.05frames/s]

100%|██████████| 3648/3648 [00:02<00:00, 1379.23frames/s]

100%|██████████| 3648/3648 [00:02<00:00, 1703.59frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/3648 [00:00<?, ?frames/s]

 80%|███████▉  | 2910/3648 [00:00<00:00, 6774.12frames/s]

100%|██████████| 3648/3648 [00:02<00:00, 1368.43frames/s]

100%|██████████| 3648/3648 [00:02<00:00, 1690.91frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/3648 [00:00<?, ?frames/s]

 80%|███████▉  | 2910/3648 [00:00<00:00, 6812.45frames/s]

100%|██████████| 3648/3648 [00:02<00:00, 1382.72frames/s]

100%|██████████| 3648/3648 [00:02<00:00, 1708.05frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/3648 [00:00<?, ?frames/s]

 80%|███████▉  | 2910/3648 [00:00<00:00, 6784.91frames/s]

100%|██████████| 3648/3648 [00:02<00:00, 1018.65frames/s]

100%|██████████| 3648/3648 [00:02<00:00, 1278.41frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/7670 [00:00<?, ?frames/s]

 33%|███▎      | 2522/7670 [00:00<00:00, 6330.52frames/s]

 71%|███████▏  | 5484/7670 [00:00<00:00, 6444.61frames/s]

 99%|█████████▊| 7568/7670 [00:01<00:00, 6249.27frames/s]

100%|██████████| 7670/7670 [00:07<00:00, 1007.50frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/7670 [00:00<?, ?frames/s]

 33%|███▎      | 2522/7670 [00:00<00:00, 6337.14frames/s]

 71%|███████▏  | 5484/7670 [00:00<00:00, 6431.27frames/s]

 99%|█████████▊| 7568/7670 [00:01<00:00, 6244.41frames/s]

100%|██████████| 7670/7670 [00:06<00:00, 1155.03frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/7670 [00:00<?, ?frames/s]

 33%|███▎      | 2522/7670 [00:00<00:00, 6358.12frames/s]

 71%|███████▏  | 5484/7670 [00:00<00:00, 6469.45frames/s]

 99%|█████████▊| 7568/7670 [00:01<00:00, 6271.79frames/s]

100%|██████████| 7670/7670 [00:07<00:00, 974.50frames/s] 


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/7670 [00:00<?, ?frames/s]

 33%|███▎      | 2522/7670 [00:00<00:00, 6339.82frames/s]

 71%|███████▏  | 5484/7670 [00:00<00:00, 6446.79frames/s]

 99%|█████████▊| 7568/7670 [00:01<00:00, 6263.43frames/s]

100%|██████████| 7670/7670 [00:05<00:00, 1393.07frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/7670 [00:00<?, ?frames/s]

 33%|███▎      | 2522/7670 [00:00<00:00, 6380.37frames/s]

 71%|███████▏  | 5484/7670 [00:00<00:00, 6479.30frames/s]

 99%|█████████▊| 7568/7670 [00:01<00:00, 6246.86frames/s]

100%|██████████| 7670/7670 [00:03<00:00, 2254.50frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/7670 [00:00<?, ?frames/s]

 33%|███▎      | 2522/7670 [00:00<00:00, 6308.24frames/s]

 71%|███████▏  | 5484/7670 [00:00<00:00, 6434.89frames/s]

 99%|█████████▊| 7568/7670 [00:01<00:00, 6261.13frames/s]

100%|██████████| 7670/7670 [00:05<00:00, 1384.69frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/7670 [00:00<?, ?frames/s]

 33%|███▎      | 2522/7670 [00:00<00:00, 6338.79frames/s]

 71%|███████▏  | 5484/7670 [00:00<00:00, 6435.27frames/s]

 99%|█████████▊| 7568/7670 [00:01<00:00, 6248.85frames/s]

100%|██████████| 7670/7670 [00:07<00:00, 1060.15frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/7670 [00:00<?, ?frames/s]

 33%|███▎      | 2522/7670 [00:00<00:00, 6317.93frames/s]

 71%|███████▏  | 5484/7670 [00:00<00:00, 6437.61frames/s]

 99%|█████████▊| 7568/7670 [00:01<00:00, 6241.93frames/s]

100%|██████████| 7670/7670 [00:06<00:00, 1203.50frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/7670 [00:00<?, ?frames/s]

 33%|███▎      | 2522/7670 [00:00<00:00, 6342.87frames/s]

 71%|███████▏  | 5484/7670 [00:00<00:00, 6438.92frames/s]

 99%|█████████▊| 7568/7670 [00:01<00:00, 6253.24frames/s]

100%|██████████| 7670/7670 [00:05<00:00, 1481.22frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/7670 [00:00<?, ?frames/s]

 33%|███▎      | 2522/7670 [00:00<00:00, 6328.40frames/s]

 71%|███████▏  | 5484/7670 [00:00<00:00, 6433.87frames/s]

 99%|█████████▊| 7568/7670 [00:01<00:00, 6247.99frames/s]

100%|██████████| 7670/7670 [00:07<00:00, 967.98frames/s] 

Audio    Duration  Avg Latency  Min Latency    RTF
--------------------------------------------------
short        6.9s       0.281s       0.278s  24.6x
medium      36.5s       2.314s       1.638s  15.8x
long        76.7s       6.428s       3.505s  11.9x


## 11. CPU Comparison & Accuracy

Run the same audio through OpenAI Whisper on CPU. Compare transcription accuracy using Word Error Rate (WER).

In [12]:
import whisper
from jiwer import wer

cpu_model = whisper.load_model("large-v3-turbo", device="cpu")

for name, info in audio_files.items():
    t0 = time.time()
    cpu_result = cpu_model.transcribe(info["path"], language="en", verbose=False)
    info["cpu_latency"] = time.time() - t0
    info["cpu_text"] = cpu_result["text"].strip()
    info["speedup"] = info["cpu_latency"] / info["neuron_avg_latency"]
    info["wer"] = wer(info["cpu_text"], info["neuron_text"])

print(f"{'Audio':<8} {'Neuron':>10} {'CPU':>10} {'Speedup':>8} {'WER':>6}")
print("-" * 46)
for name, info in audio_files.items():
    print(f"{name:<8} {info['neuron_avg_latency']:>9.3f}s {info['cpu_latency']:>9.3f}s {info['speedup']:>7.1f}x {info['wer']:>5.1%}")

print("\n--- Transcription Comparison ---")
for name, info in audio_files.items():
    match = "MATCH" if info["neuron_text"] == info["cpu_text"] else "DIFF"
    print(f"\n[{name}] ({match})")
    print(f"  Neuron: {info['neuron_text']}")
    print(f"  CPU:    {info['cpu_text']}")

  0%|                                              | 0.00/1.51G [00:00<?, ?iB/s]

  1%|▏                                    | 8.69M/1.51G [00:00<00:17, 90.7MiB/s]

  1%|▍                                    | 17.4M/1.51G [00:00<00:17, 90.9MiB/s]

  2%|▋                                    | 26.5M/1.51G [00:00<00:17, 93.2MiB/s]

  2%|▊                                    | 35.4M/1.51G [00:00<00:18, 84.5MiB/s]

  3%|█                                    | 43.6M/1.51G [00:00<00:19, 81.9MiB/s]

  3%|█▏                                   | 51.5M/1.51G [00:00<00:20, 75.4MiB/s]

  4%|█▍                                   | 61.5M/1.51G [00:00<00:18, 83.7MiB/s]

  5%|█▋                                   | 71.3M/1.51G [00:00<00:17, 89.3MiB/s]

  5%|█▉                                   | 80.0M/1.51G [00:00<00:17, 87.7MiB/s]

  6%|██                                   | 88.5M/1.51G [00:01<00:17, 88.0MiB/s]

  6%|██▎                                  | 96.9M/1.51G [00:01<00:18, 80.3MiB/s]

  7%|██▌                                   | 106M/1.51G [00:01<00:18, 83.7MiB/s]

  7%|██▊                                   | 114M/1.51G [00:01<00:18, 79.8MiB/s]

  8%|██▉                                   | 122M/1.51G [00:01<00:18, 78.6MiB/s]

  8%|███▏                                  | 129M/1.51G [00:01<00:19, 74.3MiB/s]

  9%|███▎                                  | 136M/1.51G [00:01<00:20, 73.2MiB/s]

  9%|███▌                                  | 145M/1.51G [00:01<00:19, 76.8MiB/s]

 10%|███▊                                  | 153M/1.51G [00:01<00:18, 80.6MiB/s]

 11%|████                                  | 163M/1.51G [00:02<00:16, 86.6MiB/s]

 11%|████▏                                 | 171M/1.51G [00:02<00:16, 87.3MiB/s]

 12%|████▍                                 | 180M/1.51G [00:02<00:18, 78.9MiB/s]

 12%|████▌                                 | 188M/1.51G [00:02<00:38, 37.3MiB/s]

 13%|████▊                                 | 194M/1.51G [00:02<00:34, 41.5MiB/s]

 13%|████▉                                 | 200M/1.51G [00:03<00:30, 46.3MiB/s]

 13%|█████                                 | 206M/1.51G [00:03<00:30, 45.6MiB/s]

 14%|█████▎                                | 216M/1.51G [00:03<00:23, 58.0MiB/s]

 15%|█████▌                                | 224M/1.51G [00:03<00:21, 64.1MiB/s]

 15%|█████▋                                | 233M/1.51G [00:03<00:19, 70.9MiB/s]

 16%|█████▉                                | 241M/1.51G [00:04<00:42, 32.0MiB/s]

 16%|██████                                | 248M/1.51G [00:04<00:35, 38.1MiB/s]

 17%|██████▎                               | 256M/1.51G [00:04<00:29, 45.3MiB/s]

 17%|██████▍                               | 264M/1.51G [00:04<00:25, 52.6MiB/s]

 18%|██████▋                               | 271M/1.51G [00:04<00:23, 55.7MiB/s]

 18%|██████▉                               | 279M/1.51G [00:04<00:20, 63.3MiB/s]

 19%|███████                               | 287M/1.51G [00:04<00:23, 56.9MiB/s]

 19%|███████▎                              | 296M/1.51G [00:04<00:19, 67.1MiB/s]

 20%|███████▍                              | 304M/1.51G [00:04<00:18, 70.2MiB/s]

 20%|███████▋                              | 312M/1.51G [00:05<00:17, 74.7MiB/s]

 21%|███████▉                              | 320M/1.51G [00:05<00:17, 72.5MiB/s]

 21%|████████                              | 327M/1.51G [00:05<00:23, 53.7MiB/s]

 22%|████████▏                             | 333M/1.51G [00:05<00:40, 31.1MiB/s]

 22%|████████▎                             | 339M/1.51G [00:05<00:35, 35.7MiB/s]

 22%|████████▍                             | 345M/1.51G [00:06<00:32, 39.0MiB/s]

 23%|████████▋                             | 352M/1.51G [00:06<00:26, 46.5MiB/s]

 23%|████████▊                             | 358M/1.51G [00:06<00:26, 47.0MiB/s]

 24%|████████▉                             | 364M/1.51G [00:06<00:24, 50.8MiB/s]

 24%|█████████                             | 369M/1.51G [00:06<00:23, 51.9MiB/s]

 24%|█████████▎                            | 376M/1.51G [00:06<00:21, 55.9MiB/s]

 25%|█████████▍                            | 385M/1.51G [00:06<00:18, 66.1MiB/s]

 25%|█████████▋                            | 393M/1.51G [00:06<00:17, 70.4MiB/s]

 26%|█████████▉                            | 401M/1.51G [00:06<00:16, 74.3MiB/s]

 26%|██████████                            | 409M/1.51G [00:07<00:15, 76.0MiB/s]

 27%|██████████▎                           | 417M/1.51G [00:07<00:15, 78.4MiB/s]

 28%|██████████▍                           | 426M/1.51G [00:07<00:14, 83.0MiB/s]

 28%|██████████▋                           | 434M/1.51G [00:07<00:14, 77.9MiB/s]

 29%|██████████▉                           | 443M/1.51G [00:07<00:13, 83.0MiB/s]

 29%|███████████▏                          | 453M/1.51G [00:07<00:13, 87.2MiB/s]

 30%|███████████▍                          | 462M/1.51G [00:07<00:12, 88.6MiB/s]

 31%|███████████▋                          | 472M/1.51G [00:07<00:12, 93.4MiB/s]

 31%|███████████▊                          | 481M/1.51G [00:07<00:14, 78.4MiB/s]

 32%|████████████                          | 489M/1.51G [00:08<00:13, 80.2MiB/s]

 32%|████████████▏                         | 497M/1.51G [00:08<00:13, 80.5MiB/s]

 33%|████████████▍                         | 505M/1.51G [00:08<00:15, 71.4MiB/s]

 33%|████████████▌                         | 512M/1.51G [00:08<00:15, 72.0MiB/s]

 34%|████████████▊                         | 520M/1.51G [00:08<00:14, 73.8MiB/s]

 34%|████████████▉                         | 527M/1.51G [00:08<00:14, 71.5MiB/s]

 35%|█████████████▏                        | 534M/1.51G [00:08<00:15, 68.7MiB/s]

 35%|█████████████▎                        | 541M/1.51G [00:08<00:15, 66.9MiB/s]

 36%|█████████████▌                        | 551M/1.51G [00:08<00:13, 78.9MiB/s]

 36%|█████████████▊                        | 559M/1.51G [00:09<00:13, 77.7MiB/s]

 37%|█████████████▉                        | 566M/1.51G [00:09<00:14, 72.9MiB/s]

 37%|██████████████▏                       | 577M/1.51G [00:09<00:12, 81.9MiB/s]

 38%|██████████████▍                       | 585M/1.51G [00:09<00:11, 85.1MiB/s]

 38%|██████████████▌                       | 594M/1.51G [00:09<00:11, 85.3MiB/s]

 39%|██████████████▊                       | 602M/1.51G [00:09<00:12, 76.4MiB/s]

 40%|███████████████                       | 613M/1.51G [00:09<00:11, 86.9MiB/s]

 40%|███████████████▎                      | 621M/1.51G [00:09<00:11, 80.9MiB/s]

 41%|███████████████▌                      | 629M/1.51G [00:09<00:12, 78.3MiB/s]

 41%|███████████████▋                      | 637M/1.51G [00:10<00:15, 63.2MiB/s]

 42%|███████████████▊                      | 644M/1.51G [00:10<00:18, 52.1MiB/s]

 42%|████████████████                      | 651M/1.51G [00:10<00:16, 57.1MiB/s]

 43%|████████████████▏                     | 657M/1.51G [00:10<00:19, 48.2MiB/s]

 43%|████████████████▎                     | 664M/1.51G [00:10<00:17, 53.4MiB/s]

 43%|████████████████▍                     | 670M/1.51G [00:10<00:21, 41.8MiB/s]

 44%|████████████████▌                     | 674M/1.51G [00:11<00:33, 27.3MiB/s]

 44%|████████████████▋                     | 678M/1.51G [00:11<00:31, 28.5MiB/s]

 44%|████████████████▊                     | 682M/1.51G [00:11<00:29, 30.3MiB/s]

 44%|████████████████▉                     | 685M/1.51G [00:11<00:44, 20.4MiB/s]

 45%|████████████████▉                     | 690M/1.51G [00:12<00:36, 24.6MiB/s]

 45%|█████████████████                     | 693M/1.51G [00:12<00:34, 25.5MiB/s]

 45%|█████████████████▏                    | 699M/1.51G [00:12<00:28, 31.6MiB/s]

 46%|█████████████████▎                    | 704M/1.51G [00:12<00:23, 37.6MiB/s]

 46%|█████████████████▍                    | 709M/1.51G [00:12<00:25, 34.3MiB/s]

 46%|█████████████████▌                    | 713M/1.51G [00:12<00:23, 36.9MiB/s]

 47%|█████████████████▋                    | 721M/1.51G [00:12<00:18, 47.2MiB/s]

 47%|█████████████████▉                    | 728M/1.51G [00:12<00:15, 54.4MiB/s]

 48%|██████████████████                    | 735M/1.51G [00:12<00:14, 59.2MiB/s]

 48%|██████████████████▎                   | 744M/1.51G [00:13<00:12, 68.7MiB/s]

 49%|██████████████████▍                   | 751M/1.51G [00:13<00:12, 66.9MiB/s]

 49%|██████████████████▋                   | 757M/1.51G [00:13<00:12, 63.6MiB/s]

 49%|██████████████████▊                   | 764M/1.51G [00:13<00:13, 62.4MiB/s]

 50%|███████████████████                   | 773M/1.51G [00:13<00:11, 71.5MiB/s]

 51%|███████████████████▏                  | 780M/1.51G [00:13<00:11, 71.1MiB/s]

 51%|███████████████████▍                  | 791M/1.51G [00:13<00:09, 84.4MiB/s]

 52%|███████████████████▋                  | 799M/1.51G [00:13<00:09, 81.8MiB/s]

 52%|███████████████████▉                  | 808M/1.51G [00:13<00:09, 84.9MiB/s]

 53%|████████████████████                  | 816M/1.51G [00:14<00:09, 78.7MiB/s]

 53%|████████████████████▎                 | 824M/1.51G [00:14<00:10, 73.9MiB/s]

 54%|████████████████████▍                 | 832M/1.51G [00:14<00:09, 76.0MiB/s]

 54%|████████████████████▋                 | 839M/1.51G [00:14<00:10, 72.5MiB/s]

 55%|████████████████████▊                 | 846M/1.51G [00:14<00:10, 71.3MiB/s]

 55%|█████████████████████                 | 854M/1.51G [00:14<00:09, 74.4MiB/s]

 56%|█████████████████████▏                | 861M/1.51G [00:14<00:09, 71.9MiB/s]

 56%|█████████████████████▍                | 869M/1.51G [00:14<00:09, 73.8MiB/s]

 57%|█████████████████████▌                | 876M/1.51G [00:14<00:09, 73.1MiB/s]

 57%|█████████████████████▋                | 883M/1.51G [00:14<00:09, 72.6MiB/s]

 58%|█████████████████████▉                | 892M/1.51G [00:15<00:08, 77.1MiB/s]

 58%|██████████████████████▏               | 900M/1.51G [00:15<00:08, 79.9MiB/s]

 59%|██████████████████████▎               | 907M/1.51G [00:15<00:08, 78.5MiB/s]

 59%|██████████████████████▌               | 915M/1.51G [00:15<00:08, 77.1MiB/s]

 60%|██████████████████████▊               | 925M/1.51G [00:15<00:07, 86.7MiB/s]

 61%|██████████████████████▉               | 934M/1.51G [00:15<00:07, 80.2MiB/s]

 61%|███████████████████████▏              | 944M/1.51G [00:15<00:07, 86.9MiB/s]

 62%|███████████████████████▍              | 952M/1.51G [00:15<00:07, 85.7MiB/s]

 62%|███████████████████████▋              | 960M/1.51G [00:15<00:07, 82.4MiB/s]

 63%|███████████████████████▊              | 968M/1.51G [00:16<00:07, 79.9MiB/s]

 63%|████████████████████████              | 976M/1.51G [00:16<00:11, 53.7MiB/s]

 64%|████████████████████████▍             | 990M/1.51G [00:16<00:07, 74.1MiB/s]

 65%|████████████████████████▌             | 999M/1.51G [00:16<00:07, 78.7MiB/s]

 65%|████████████████████████▏            | 0.98G/1.51G [00:16<00:06, 80.3MiB/s]

 66%|████████████████████████▍            | 0.99G/1.51G [00:16<00:08, 68.2MiB/s]

 66%|████████████████████████▌            | 1.00G/1.51G [00:16<00:07, 72.4MiB/s]

 67%|████████████████████████▊            | 1.01G/1.51G [00:17<00:07, 73.4MiB/s]

 67%|████████████████████████▉            | 1.02G/1.51G [00:17<00:07, 75.0MiB/s]

 68%|█████████████████████████▏           | 1.02G/1.51G [00:17<00:06, 77.0MiB/s]

 68%|█████████████████████████▎           | 1.03G/1.51G [00:17<00:06, 78.4MiB/s]

 69%|█████████████████████████▌           | 1.04G/1.51G [00:17<00:06, 71.8MiB/s]

 69%|█████████████████████████▋           | 1.05G/1.51G [00:17<00:07, 69.1MiB/s]

 70%|█████████████████████████▊           | 1.05G/1.51G [00:17<00:06, 71.1MiB/s]

 70%|██████████████████████████           | 1.06G/1.51G [00:17<00:06, 75.2MiB/s]

 71%|██████████████████████████▏          | 1.07G/1.51G [00:17<00:06, 75.4MiB/s]

 72%|██████████████████████████▍          | 1.08G/1.51G [00:18<00:05, 83.2MiB/s]

 72%|██████████████████████████▋          | 1.09G/1.51G [00:18<00:05, 76.0MiB/s]

 73%|██████████████████████████▊          | 1.09G/1.51G [00:18<00:07, 63.5MiB/s]

 73%|██████████████████████████▉          | 1.10G/1.51G [00:18<00:07, 62.2MiB/s]

 73%|███████████████████████████▏         | 1.11G/1.51G [00:18<00:07, 59.6MiB/s]

 74%|███████████████████████████▎         | 1.11G/1.51G [00:18<00:06, 65.6MiB/s]

 74%|███████████████████████████▍         | 1.12G/1.51G [00:18<00:06, 65.5MiB/s]

 75%|███████████████████████████▋         | 1.13G/1.51G [00:18<00:06, 67.3MiB/s]

 75%|███████████████████████████▊         | 1.13G/1.51G [00:18<00:05, 70.3MiB/s]

 76%|████████████████████████████         | 1.14G/1.51G [00:19<00:05, 67.6MiB/s]

 76%|████████████████████████████▏        | 1.15G/1.51G [00:19<00:05, 66.2MiB/s]

 77%|████████████████████████████▎        | 1.15G/1.51G [00:19<00:11, 33.2MiB/s]

 77%|████████████████████████████▌        | 1.16G/1.51G [00:19<00:08, 42.5MiB/s]

 78%|████████████████████████████▋        | 1.17G/1.51G [00:19<00:07, 49.9MiB/s]

 78%|████████████████████████████▉        | 1.18G/1.51G [00:19<00:06, 56.7MiB/s]

 79%|█████████████████████████████        | 1.18G/1.51G [00:20<00:05, 59.5MiB/s]

 79%|█████████████████████████████▎       | 1.19G/1.51G [00:20<00:05, 64.4MiB/s]

 80%|█████████████████████████████▍       | 1.20G/1.51G [00:20<00:04, 66.5MiB/s]

 80%|█████████████████████████████▌       | 1.21G/1.51G [00:20<00:04, 69.5MiB/s]

 80%|█████████████████████████████▊       | 1.21G/1.51G [00:20<00:04, 69.7MiB/s]

 81%|█████████████████████████████▉       | 1.22G/1.51G [00:20<00:04, 71.3MiB/s]

 81%|██████████████████████████████▏      | 1.23G/1.51G [00:20<00:04, 73.9MiB/s]

 82%|██████████████████████████████▎      | 1.24G/1.51G [00:20<00:03, 75.6MiB/s]

 82%|██████████████████████████████▌      | 1.24G/1.51G [00:20<00:03, 75.5MiB/s]

 83%|██████████████████████████████▋      | 1.25G/1.51G [00:20<00:03, 77.9MiB/s]

 83%|██████████████████████████████▉      | 1.26G/1.51G [00:21<00:03, 77.6MiB/s]

 84%|███████████████████████████████      | 1.27G/1.51G [00:21<00:03, 73.0MiB/s]

 84%|███████████████████████████████▏     | 1.27G/1.51G [00:21<00:03, 72.7MiB/s]

 85%|███████████████████████████████▍     | 1.28G/1.51G [00:21<00:03, 80.5MiB/s]

 86%|███████████████████████████████▋     | 1.29G/1.51G [00:21<00:02, 82.9MiB/s]

 86%|███████████████████████████████▊     | 1.30G/1.51G [00:21<00:02, 82.8MiB/s]

 87%|████████████████████████████████     | 1.31G/1.51G [00:21<00:02, 84.2MiB/s]

 87%|████████████████████████████████▎    | 1.31G/1.51G [00:21<00:02, 78.6MiB/s]

 88%|████████████████████████████████▍    | 1.32G/1.51G [00:21<00:02, 79.6MiB/s]

 88%|████████████████████████████████▌    | 1.33G/1.51G [00:22<00:02, 65.0MiB/s]

 89%|████████████████████████████████▊    | 1.34G/1.51G [00:22<00:02, 70.1MiB/s]

 89%|████████████████████████████████▉    | 1.34G/1.51G [00:22<00:02, 72.1MiB/s]

 90%|█████████████████████████████████▏   | 1.35G/1.51G [00:22<00:02, 75.5MiB/s]

 90%|█████████████████████████████████▎   | 1.36G/1.51G [00:22<00:02, 72.9MiB/s]

 91%|█████████████████████████████████▌   | 1.37G/1.51G [00:22<00:02, 63.1MiB/s]

 91%|█████████████████████████████████▋   | 1.37G/1.51G [00:22<00:02, 63.1MiB/s]

 91%|█████████████████████████████████▊   | 1.38G/1.51G [00:22<00:02, 58.7MiB/s]

 92%|█████████████████████████████████▉   | 1.38G/1.51G [00:23<00:02, 52.8MiB/s]

 92%|██████████████████████████████████▏  | 1.39G/1.51G [00:23<00:02, 58.1MiB/s]

 93%|██████████████████████████████████▎  | 1.40G/1.51G [00:23<00:01, 64.6MiB/s]

 93%|██████████████████████████████████▍  | 1.40G/1.51G [00:23<00:03, 32.9MiB/s]

 94%|██████████████████████████████████▋  | 1.41G/1.51G [00:23<00:02, 41.6MiB/s]

 94%|██████████████████████████████████▊  | 1.42G/1.51G [00:23<00:02, 45.1MiB/s]

 95%|██████████████████████████████████▉  | 1.42G/1.51G [00:24<00:01, 49.3MiB/s]

 95%|███████████████████████████████████▏ | 1.43G/1.51G [00:24<00:01, 57.8MiB/s]

 96%|███████████████████████████████████▎ | 1.44G/1.51G [00:24<00:01, 60.1MiB/s]

 96%|███████████████████████████████████▌ | 1.45G/1.51G [00:24<00:01, 63.2MiB/s]

 96%|███████████████████████████████████▋ | 1.45G/1.51G [00:24<00:00, 67.3MiB/s]

 97%|███████████████████████████████████▊ | 1.46G/1.51G [00:24<00:00, 70.3MiB/s]

 97%|████████████████████████████████████ | 1.47G/1.51G [00:24<00:00, 51.3MiB/s]

 98%|████████████████████████████████████▏| 1.48G/1.51G [00:24<00:00, 59.6MiB/s]

 98%|████████████████████████████████████▍| 1.48G/1.51G [00:25<00:00, 65.7MiB/s]

 99%|████████████████████████████████████▌| 1.49G/1.51G [00:25<00:00, 67.3MiB/s]

 99%|████████████████████████████████████▊| 1.50G/1.51G [00:25<00:00, 35.5MiB/s]

100%|████████████████████████████████████▉| 1.50G/1.51G [00:25<00:00, 38.9MiB/s]

100%|█████████████████████████████████████| 1.51G/1.51G [00:25<00:00, 62.9MiB/s]

/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/691 [00:00<?, ?frames/s]

100%|██████████| 691/691 [00:16<00:00, 42.38frames/s]

100%|██████████| 691/691 [00:16<00:00, 42.38frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/3648 [00:00<?, ?frames/s]

 80%|███████▉  | 2912/3648 [00:14<00:03, 195.64frames/s]

100%|██████████| 3648/3648 [00:27<00:00, 118.39frames/s]

100%|██████████| 3648/3648 [00:27<00:00, 130.75frames/s]


/opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/lib/python3.12/site-packages/whisper/transcribe.py:132: UserWarning: FP16 is not supported on CPU; using FP32 instead
  warnings.warn("FP16 is not supported on CPU; using FP32 instead")


  0%|          | 0/7670 [00:00<?, ?frames/s]

 33%|███▎      | 2522/7670 [00:14<00:29, 172.76frames/s]

 71%|███████▏  | 5484/7670 [00:29<00:11, 185.53frames/s]

 71%|███████▏  | 5484/7670 [00:41<00:11, 185.53frames/s]

 99%|█████████▊| 7568/7670 [00:44<00:00, 166.91frames/s]

 99%|█████████▊| 7568/7670 [01:01<00:00, 166.91frames/s]

100%|██████████| 7670/7670 [02:35<00:00, 29.06frames/s] 

100%|██████████| 7670/7670 [02:35<00:00, 49.45frames/s]

Audio        Neuron        CPU  Speedup    WER
----------------------------------------------
short        0.281s    16.366s    58.2x  0.0%
medium       2.314s    27.984s    12.1x  2.9%
long         6.428s   155.220s    24.1x 25.8%

--- Transcription Comparison ---

[short] (MATCH)
  Neuron: The quick brown fox jumps over the lazy dog. This is a test of automatic speech recognition.
  CPU:    The quick brown fox jumps over the lazy dog. This is a test of automatic speech recognition.

[medium] (DIFF)
  Neuron: Artificial intelligence has transformed the way we interact with technology. Machine learning models can now understand and generate human language with remarkable accuracy. Speech recognition systems like Whisper have made it possible to transcribe audio in real time. Supporting dozens of languages and handling noisy environments with ease. The combination of hardware acceleration and optimized software makes these capabilities accessible at scale. Enabling applications from liv

## 12. Results Summary

In [13]:
# Collect all results into a summary dict
results_summary = {
    "platform": PLATFORM,
    "model": "openai/whisper-large-v3-turbo",
    "precision": "FP16",
    "tp_degree": 1,
    "compile_time_s": compile_time,
    "load_time_s": round(load_time, 1),
    "benchmarks": {},
}

for name, info in audio_files.items():
    results_summary["benchmarks"][name] = {
        "audio_duration_s": round(info["duration"], 1),
        "neuron_avg_latency_s": round(info["neuron_avg_latency"], 3),
        "neuron_min_latency_s": round(info["neuron_min_latency"], 3),
        "rtf": round(info["neuron_rtf"], 1),
        "cpu_latency_s": round(info["cpu_latency"], 3),
        "speedup_vs_cpu": round(info["speedup"], 1),
        "wer_vs_cpu": round(info["wer"], 4),
        "neuron_text": info["neuron_text"],
        "cpu_text": info["cpu_text"],
    }

# Save to JSON
results_file = f"/tmp/whisper_results_{PLATFORM}.json"
with open(results_file, "w") as f:
    json.dump(results_summary, f, indent=2)
print(f"Results saved to {results_file}")
print(json.dumps(results_summary, indent=2))

Results saved to /tmp/whisper_results_inf2.json
{
  "platform": "inf2",
  "model": "openai/whisper-large-v3-turbo",
  "precision": "FP16",
  "tp_degree": 1,
  "compile_time_s": 38.1443510055542,
  "load_time_s": 11.4,
  "benchmarks": {
    "short": {
      "audio_duration_s": 6.9,
      "neuron_avg_latency_s": 0.281,
      "neuron_min_latency_s": 0.278,
      "rtf": 24.6,
      "cpu_latency_s": 16.366,
      "speedup_vs_cpu": 58.2,
      "wer_vs_cpu": 0.0,
      "neuron_text": "The quick brown fox jumps over the lazy dog. This is a test of automatic speech recognition.",
      "cpu_text": "The quick brown fox jumps over the lazy dog. This is a test of automatic speech recognition."
    },
    "medium": {
      "audio_duration_s": 36.5,
      "neuron_avg_latency_s": 2.314,
      "neuron_min_latency_s": 1.638,
      "rtf": 15.8,
      "cpu_latency_s": 27.984,
      "speedup_vs_cpu": 12.1,
      "wer_vs_cpu": 0.0286,
      "neuron_text": "Artificial intelligence has transformed the way we

## 13. Transcribe Your Own Audio

Replace the path below with any audio file (WAV, MP3, FLAC, etc.).

In [14]:
# my_audio = "/path/to/your/audio.wav"
# result = model.transcribe(my_audio, language="en", verbose=True)
# print(result["text"])

## Performance Results

*(Results will be filled in after testing on both platforms)*

### trn2.3xlarge (LNC=2, SDK 2.28)

| Audio | Duration | Neuron Latency | RTF | CPU Latency | Speedup | WER |
|-------|----------|---------------|-----|-------------|---------|-----|
| short | | | | | | |
| medium | | | | | | |
| long | | | | | | |

- Compilation: s
- Model load: s

### inf2.xlarge (SDK 2.28)

| Audio | Duration | Neuron Latency | RTF | CPU Latency | Speedup | WER |
|-------|----------|---------------|-----|-------------|---------|-----|
| short | | | | | | |
| medium | | | | | | |
| long | | | | | | |

- Compilation: s
- Model load: s

### Optimizations Applied

| Optimization | Description | Impact |
|---|---|---|
| Cross-attn K/V cache | Skip redundant K/V projections during decode | ~2.5x decode latency reduction |
| Fused QKV | Single matmul for Q/K/V in self-attention | Reduced kernel launch overhead |
| NKI flash attention | Hardware flash attention in 32 encoder layers | 45% faster compilation |
| NKI fused Conv1D+GELU | Fused encoder frontend convolutions | Marginal (encoder frontend not bottleneck) |

All NKI kernels are from existing libraries (NxDI SDK and nki-library) — no custom kernels.